### ISSUE TITLE
hical not falling back to ZeroDark when ZeroDarkRate is enabled but matrix file is not available.

### ISSUE THREAD

**Issue by rbeyer** [DOI-USGS/ISIS3#4931]
**ISIS version(s) affected**: 6.0.0

**Description**  
When the new ZeroDarkRate module is enabled for ISIS `hical` it uses a set of `DarkRate*.csv` files in $ISISDATA/mro/calibration/matrices for calculations.  These files have names like `DarkRate_RED2_1_TDI128_BIN4_ADC54_hical_0002.csv` and there could be a file for each combination of CCD, channel, TDI rate, binning value, and ADC setting (which is a potentially open-ended and infinite number of combinations).

However, if `hical` attempts to process an image for which the appropriate `DarkRate*.csv` does not exist, it errors.

**How to reproduce**  
1. First, set up appropriate .conf file to enable the ZeroDarkRate module, `hical-darktest.conf`.
2. Download http://hirise-pds.lpl.arizona.edu/PDS/EDR/ESP/ORB_020200_020299/ESP_020242_0945/ESP_020242_0945_RED0_1.IMG
3. Then:

```
$> hi2isis fr= ESP_020242_0945_RED0_1.IMG to= ESP_020242_0945_RED0_1.cub
$> hical conf= hical-darktest.conf fr= ESP_020242_0945_RED0_1.cub to= ESP_020242_0945_RED0_1.hical.cub
**USER ERROR** Failed in ZeroDarkRate module.
**ERROR** No existing files found with a numerial version matching [DarkRate_RED0_1_TDI128_BIN4_ADC54_hical_????.csv] in [/Users/rbeyer/opt/isisdata/mro/calibration/matrices].
```

**Possible Solution**  
When ZeroDarkRate is enabled and no appropriate `DarkRate*.csv` file can be found, fall back to ZeroDark behavior.

Alternately, leave as-is, since it is a legit error, given the specification, and leave the remedy up to the user to use just the ZeroDark behavior.

**Additional context**
Not all possible combinations of `DarkRate*.csv` exist because there is insufficient calibration data to create them.

**Comment by AustinSanders** [comment]
I haven't yet replicated this issue, but I'm tracking this issue and plan to address this (or at least triage) with continuous support hours.

**Comment by jlaura** [comment]
We discussed this a bit yesterday in the standup. While the auto-fail back might be nice for some users, I was worried that one might process a hunk of data and then have to know to go and look at the individual labels to know what dark file files had been used. It seems that a more transparent way to have this work would be to throw an error when the ZeroDarkRate can not find the appropriate CSV and then give the user the command to run in fall back mode. I am not a user of these applications so it may be that a less visible fall back is desirable.

**Comment by jessemapel** [comment]
The fundamental question is: Would you rather have some of the images in your data set be calibrated with ZeroDark and some with ZeroDarkRate, or all of them calibrated with ZeroDark?

**Comment by rbeyer** [comment]
These are great questions, and I wish we had talked more about them before too much work had been done.  I've thought about this some more, and of the two remedies I proposed, I think that "erroring" and then providing a more detailed error message is the better solution.  The existing error message is very good, but maybe add a second sentence:

> Not all combinations of CCD, channel, TDI rate, binning value, and ADC setting have a DarkRate*.csv available, and you may need to run hical with ZeroDark instead of ZeroDarkRate specified in the configuration file.

But maybe this is too long?

**Comment by AustinSanders** [comment]
There's a comment on #4986 about potentially adding a "--fallback" flag that could be passed to hical to enable the program to fall back to ZeroDark if ZeroDarkRate fails.  By default, the program would simply error and print the message, and with the flag, it would fall back to ZeroDark.  Is this worth implementing, or would we prefer to stick with simply adding a more informative error message?

**Comment by jessemapel** [comment]
If it's not too much work, I think it would be worth it. It should be an option on the ZeroDarkRate configuration, and not a command line option.

If it takes any significant work, I don't think it's worth it.

**Comment by audiefen** [comment]
From the instrument team perspective, it is not desirable to have our pipelines failing and then us having to do some kind of manual intervention.

If it isn't too much work to add the "--fallback flag", that would be great.

Alternatively, we could train our pipelines to use the old ZeroDark configuration when processing imaging modes where there are no existing ZeroDarkRate files. We know which imaging modes have such files and which don't.

**Comment by rbeyer** [comment]
Right, so I think Audrie and I are in support of Jesse's suggestion:

> If it's not too much work, I think it would be worth it. It should be an option on the ZeroDarkRate configuration, and not a command line option.

In the HiRISE pipelines we can build logic to deal with a ZDR HiCal run, and then run it with just ZD, but it would be less work for us if we could just add a flag or parameter in the configuration file so that ISIS could do that for us.

Yes, a command-line option that is only applicable to one setting in a configuration file isn't the right arrangement, but the concept would be wonderful, and specifying it in the ZDR configuration seems appropriate, if possible.  I'm not sure if that implies that you would then need to then have a ZD Group "inside" the ZDR group or if there's simply a way to have the "Debug::SkipModule" in ZDR set to "fallback" instead of True or False, and then the ZD group would have to be set to "true"?  I'll let you guys figure all that out, and maybe if that ramps up into "hard" then we bail?

**Comment by AustinSanders** [comment]
#4986 has been updated with the ability to fall back to ZeroDark by specifying Fallback = True within the ZeroDarkRate profile in the config file.  Any feedback on the error messages would be appreciated.

In order to avoid breaking existing pipelines, "Fallback" is not required in the profile.  If "Fallback" is not present in the config, it defaults to False, and the new functionality will not trigger.  Instead, the user will just see an improved error message and the program will error out.

**Comment by AustinSanders** [comment]
Closed via #4986
